src/extension.ts

```diff
@@ -2841,3 +2814,36 @@ function _hide_skip_taskbar_windows() {
             default_getwindowlist_windowswitcher;
     }
 }
+
+/**
+ * Moved skip task bar checking on this function/method
+ * Synchronized with ShellTracker type checks and watch out for attached dialogs
+ *
+ * Thanks to Bananaman and upstream gnome-shell devs for the information
+ *
+ * https://github.com/pop-os/shell/issues/1251
+ */
+function is_valid_minimize_to_tray(meta_win: Meta.Window, ext: Ext) {
+    let cfg = ext.conf;
+    let valid_min_to_tray = false;
+    switch (meta_win.window_type) {
+        case Meta.WindowType.NORMAL:
+        case Meta.WindowType.DIALOG:
+        case Meta.WindowType.UTILITY:
+        case Meta.WindowType.MODAL_DIALOG:
+            valid_min_to_tray = true;
```

Comment:
I can add this is the repro of OR windows the break my shell script? Sorry to throw it back to you.

Reply:
@jmmaranan The break-my-pop-shell script shows one kind of OR window (a "skip-taskbar" window with override-redirect (OR) enabled). But there are other kinds.

The thing that GNOME Shell does internally, is:

- Never track OR windows, no matter what window type they are.
- Filter remaining (non-OR) windows to only keep NORMAL, DIALOG, UTILITY and MODAL_DIALOG.

So to reproduce exact behavior we need to check `is_override_redirect()` too, just like they do. :)

Reply:
Maybe Android dev windows by Jetbrains or Eclipse would challenge this part of the code.

Reply:
@jmmaranan I don't think so. Override-redirect windows are filtered by Gnome already (because they are internal windows that are not allowed to be window-managed), so I don't think that Jetbrains/Eclipse uses OR for any of its normal windows.

From the GNOME Shell discussions they made it clear that GNOME never tracks OR windows for any reason, because it makes no sense. The other "allowed window types" filters just further narrowed down the list of non-OR windows.

I suggest testing the proposed `valid_min_to_tray = !meta_win.is_override_redirect();` with those IDEs. I would be very surprised if it causes any problems. Because removing OR windows is correct. We've never been able to display OR windows in the GNOME Overview. It would crash the shell. :)

Reply:
I did a little debugging on this. I think for the purposes of the Overview handling and Minimized to Tray - the logic should only use the NORMAL and UTILITY or just NORMAL. The Save/Save As or Quit dialogs are well of course DIALOGs and they will show on the overview if not excluded (based on your comment about not having them show on the App/Tab Switchers, the comment was put there but looks like it is more of a comment for the Overview/Workspace not the Switchers - I already found the issue/fix for the switchers showing duplicate of the same parent icon). And finally, for some reason, the parent and the modal/dialog when debugging are tracked as a single instance by TabList and it is hard to distinguish by just asking if it is an attached dialog.

Reply:
@jmmaranan Hmm, that's some awesome testing and a very good point about the weird window types! Yeah, we should probably not handle every weird little window that GNOME handles.

Excluding nonsense windows (like modal dialogs) makes sense. And I also think that we shouldn't worry that some apps may be misclassifying their dialogs, because the only correct one for attached child dialogs is `MODAL_DIALOG`, so it's correct to exclude that one. If any application misbehaves, it's not our fault and should be patched in their app instead. :)

So, `NORMAL` and _possibly_ also `DIALOG` sounds like a perfect list for the "show minimized" feature.

One thing that comes to mind when we think about this, is that there are apparently _some_ applications that **only** create a `DIALOG` without a parent, and they use that `DIALOG` as their only (main) window. I don't know any examples that do that, I've just heard about it in the GNOME discussions (on their ticket/merge request).

But that seems extremely rare. So I honestly think excluding `DIALOG` too makes sense, and only keeping `NORMAL`. Just to avoid the risk of a bunch of weird non-modal dialogs (used for stuff like floating toolbars / color palette / painting tools etc) appearing in the Overview.

Hmm... perhaps I have an idea how this may be combined to get the proper behavior...

- Be sure that GNOME handles the main "should this window be included in overview?" test the way it normally would without Pop Shell. So if an unmodiifed GNOME would have shown the window, then it should be shown even when Pop Shell is enabled. Because their code takes very good care of things like modal dialogs attached to parents (it perfectly only shows the parent, with an attached dialog image on top of it). And normal GNOME also takes care of properly showing unattached/solo `DIALOG`-only apps. So if the core GNOME "what to show in Overview" behavior is kept as the 1st check, then the normal functionality of Overview will work properly.
- In fact, I think Pop Shell already does exactly that, and only extends the GNOME Overview list by including _extra_ windows?
- And then... if the window fails GNOME's check, Pop Shell would finally do its own "if window has skip-taskbar (meaning minimized), is not an `is_override_redirect()` window, and is a `NORMAL` window, add it to the overview too".
- That way Pop Shell will safely add _only_ the minimized windows that have a `NORMAL` type and aren't OR windows.
- This would be very robust and takes care of all normal applications.
- We could also try including `DIALOG` in Pop Shell's list, but we might see duplicate/weird windows from doing that. And it might not be useful enough to be worth that risk. It would only be necessary if an app uses a `DIALOG`-only main window and is minimized to taskbar, which is a strange combo! :)

What are your thoughts? Could this work?

Reply:
> In fact, I think Pop Shell already does exactly that, and only extends the existing GNOME Overview list by including extra windows?

Yeah @Bananaman - that's what this (Include NORMAL + MINIMIZE to TRAY WINDOWS) feature is trying to do. I think `UTILITY` makes more sense to include with `NORMAL`. One app that I have tested is GIMP if using Non-Single Window mode that fits this. The minimum checklist I guess is: 

- to include the NORMAL window that's a skip task bar
- AND not an override redirect
- AND not Gjs nor Gnome-shell WM_CLASS
- AND it isn't an attached dialog (maybe already excluded by not including WindowType's DIALOG or MODAL_DIALOG)
- AND it isn't added as a Skip Taskbar exception
- BUT it is a UTILITY Window (acting like a separate window from parent but under the same process?)

Reply:
@jmmaranan Ahh, interesting to hear about GIMP! I was wondering what `UTILITY` windows were for! Thanks for your thorough testing, it's really great to see! :)

Your checklist looks great!

> AND it isn't an attached dialog (maybe already excluded by not including WindowType's DIALOG or MODAL_DIALOG)

I would suspect that, yeah. It's probably impossible to attach a normal window to a window. It probably has to be a `MODAL_DIALOG` (and maybe also `DIALOG`) to be able to attach to another window as a focus-stealing dialog. So I think excluding those types would automatically remove all attached dialogs.

But I also suppose that adding a simple check for `!is_attached_dialog()` is fast and harmless anyway? We could include that check since we aren't sure if it's needed or not. What's your thought on this?

> AND it isn't added as a Skip Taskbar exception

What's a skip-taskbar exception?

> AND it isn't added as a Skip Taskbar exception
> BUT it is a UTILITY Window (acting like a separate window from parent but under the same process?)

I guess the two bullets above were meant to be read together. I think you're saying "Also include UTILITY window if it isn't a skip-taskbar exception"?

I don't understand this part of the list yet, but since everything else made sense, this idea probably makes sense too! :D

Sounds good!

Reply:
A `skip_task_bar exception` is a skip task bar enabled window but don't want it to be handled. As Michael mentioned that semi-configurable (somebody has to create a PR to update) ignore list in Pop-Shell on one of the earlier comments on your ticket 1251.

> I guess the two bullets above were meant to be read together. I think you're saying "Also include UTILITY window if it isn't a skip-taskbar exception"?

I was writing the list as pseudo-code :) and that last line has to begin with `AND ...`. All of them or nothing.

Reply:
@jmmaranan Ahhh. So it's a pop shell feature to exclude some minimized windows from this feature? Okay gotcha. :)

> I was writing the list as pseudo-code :)

Oh, yeah, I see that now. :) It looks correct!